pkg/reconciler/apis/apibinding/apibinding_reconcile_permissionclaim.go

```diff
@@ -0,0 +1,302 @@
+limitations under the License.
+*/
+
+package apibinding
+
+import (
+	"context"
+	"fmt"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/kcp-dev/logicalcluster"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/informers"
+	"k8s.io/klog/v2"
+
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1/permissionclaims"
+	conditionsv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/apis/conditions/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/apis/third_party/conditions/util/conditions"
+)
+
+type permissionClaimHelper struct {
+	claim apisv1alpha1.PermissionClaim
+	label string
+	key   string
+
+	lister informers.GenericInformer
+}
+
+func (p permissionClaimHelper) String() string {
+	return p.claim.String()
+}
+
+type mapGVRToClaim map[schema.GroupVersionResource][]permissionClaimHelper
+
+func (m mapGVRToClaim) String() string {
+	s := "["
+	for _, claims := range m {
+		s += fmt.Sprintf("%s", claims)
+	}
+	s = s + "]"
+	return s
+}
+
+func (c *controller) createPermissionClaimHelper(pc apisv1alpha1.PermissionClaim) (permissionClaimHelper, schema.GroupVersionResource, error) {
+
+	// get informer for the object.
+	lister, gvr, err := c.getInformerForGroupResource(pc.Group, pc.Resource)
+	if err != nil {
+		return permissionClaimHelper{}, gvr, err
+	}
+
+	key, label, err := permissionclaims.PermissionClaimToLabel(pc)
+	if err != nil {
+		return permissionClaimHelper{}, gvr, err
+	}
+
+	return permissionClaimHelper{
+		claim:  pc,
+		label:  label,
+		key:    key,
+		lister: lister,
+	}, gvr, nil
+}
+
+// reconcilePermissionClaims will determine the resources that need to be labeled for access by a permission claim.
+// It will determine what permssions need to be added, what permissions need to be removed.
+// It will also update the status if it finds an invalid permission claim.
+// Permission claims are considered invalid when the identity hashes are mismatched, and when there is no dynamic informer
+// for the group resource.
+func (c *controller) reconcilePermissionClaims(ctx context.Context, apiBinding *apisv1alpha1.APIBinding) error {
+	identityMismatch := false
+	lc := logicalcluster.From(apiBinding)
+
+	// Get all the bound resources in the locicalcluster
+	bindings, err := c.listAPIBindings(lc)
+	if err != nil {
+		return err
+	}
+	grsToBoundResource := map[apisv1alpha1.GroupResource]apisv1alpha1.BoundAPIResource{}
+	for _, binding := range bindings {
+		for _, resource := range binding.Status.BoundResources {
+			grsToBoundResource[apisv1alpha1.GroupResource{Group: resource.Group, Resource: resource.Resource}] = resource
```

Comment:
Note to self: the intent here is to ensure that a permission claim is referring to an API that actually exists in this logical cluster - if someone tries to claim some other identity, it cannot be valid.

We probably want to reconsider what happens when some but not all APIBindings fail to accept a particular permission claim, so the service provider can do something intelligent here